### ISSUE TITLE
Bug: 

### ISSUE THREAD

**Issue by maybemaynot** [pixijs/pixijs#9117]
### Current Behavior

Text stroke line of top side missed when strokeThickness = 1 or 2,
means that the bottom line always greater than top line by 2 pixels

### Expected Behavior

Text stroke line should be same weight on all the 4 side

### Steps to Reproduce

  const view = document.querySelector("canvas");
  const width = 1280;
  const height = 720;

  const background = "#000";
  const app = new PIXI.Application({ view, width, height, background });

  const style01 = new PIXI.TextStyle({
    fontFamily: "Arial",
    fontSize: 256,
    fill: "#fff",
    stroke: "#f00",
    strokeThickness: 1,
    lineJoin: "round",
  });
  const layer01 = new PIXI.Text("T", style01);
  layer01.anchor.set(0.5);
  layer01.position.set(640, 360);

  app.stage.addChild(layer01);

### Environment

- **`pixi.js` version**: *e.g. 7.1.0*
- **Browser & Version**: *e.g. Chrome 108*
- **OS & Version**: *e.g. Ubuntu 22.04*
- **Running Example**: *e.g. <https://pixiplayground.com/>*

My code on 109.0.5414.120（正式版本） （64 位）

### Possible Solution

I think that "put the fill code block before of the stroke code in Text.ts" may be a good idea ?

the code of Text.ts line 315~332 is:
```
                if (style.stroke && style.strokeThickness)
                {
                    this.drawLetterSpacing(
                        lines[i],
                        linePositionX + style.padding,
                        linePositionY + style.padding - dsOffsetText,
                        true
                    );
                }

                if (style.fill)
                {
                    this.drawLetterSpacing(
                        lines[i],
                        linePositionX + style.padding,
                        linePositionY + style.padding - dsOffsetText
                    );
                }
```

**Comment by SuperSodaSea** [comment]
This is caused by not setting the `resolution`, therefore, you will not get pixel-perfect results on systems with high DPI. Try to add `resolution: window.devicePixelRatio` to the options when you create the `PIXI.Application`.
Try the demo here: <https://pixiplayground.com/#/edit/Eec7wJ6epp8i2GMkTlywE>

**Comment by maybemaynot** [comment]
> This is caused by not setting the `resolution`, therefore, you will not get pixel-perfect results on systems with high DPI. Try to add `resolution: window.devicePixelRatio` to the options when you create the `PIXI.Application`. Try the demo here: https://pixiplayground.com/#/edit/Eec7wJ6epp8i2GMkTlywE

add resolution change noting!

  const app = new PIXI.Application({
    view,
    width,
    height,
    background,
    resolution: window.devicePixelRatio,
  });

the view like this

![image](https://user-images.githubusercontent.com/88305215/215314557-da046bed-a6b5-4950-bae8-22608b7f1333.png)

**Comment by maybemaynot** [comment]
the canvas draw result ard different between "stroke before fill" and "fill before stroke"

this is stroke before fill
![image](https://user-images.githubusercontent.com/88305215/215315096-060d87d4-6243-4237-841b-428303a3fbff.png)

and bellow is fill before stroke

**Comment by SuperSodaSea** [comment]
It make sense to put fill before stroke.
By the way, could you try adding one more option `autoDensity: true` to `PIXI.Application` and see if it works?

**Comment by maybemaynot** [comment]
> It make sense to put fill before stroke. By the way, could you try adding one more option `autoDensity: true` to `PIXI.Application` and see if it works?

nothing happen

**Comment by SuperSodaSea** [comment]
Okay, I'll fix that later.

**Comment by maybemaynot** [comment]
> Okay, I'll fix that later.

wait a minute...

I think the order of fill and stroke is not the main issue .
Even if fill after stroke, the border (eg: stroke line) of top side and bottom side should be 1pixel both, when strokeThickness = 2
But now, as my result through pixi.js version 7.1.1, the top border is 1pixel, the bottom border is 3 pixel

as i see
the actual stroke line in canvas context maybe half in text body, half out of text

In other words, this case more like border-box & content-box of CSS box-sizing
stroke + fill ~ content-box ----- means text body weight not changed when thickness changed
fill + stroke ~ border-box ----- means text body weight changed when thickness change

in my case(content box), the **REAL BUG** is the line weight between top side and bottom side of the text **IS NOT EQUAL**
that's all

**Comment by maybemaynot** [comment]
> It make sense to put fill before stroke. By the way, could you try adding one more option `autoDensity: true` to `PIXI.Application` and see if it works?

About the order of fill and stroke
Should we start another issue ?

maybe add an option to TextStyle? to switch the two case

fill + stroke (part of fill are overwriten by stroke line with inner round corner)
![image](https://user-images.githubusercontent.com/88305215/215318766-de718f92-b6df-4f17-843c-dcda26f7a207.png)

stroke + fill (half of stroke be overwriten by fill, and inner corner are not round)
![image](https://user-images.githubusercontent.com/88305215/215318778-6a40abba-f1b7-4721-9f8f-f6097404a8ba.png)

**Comment by bigtimebuddy** [comment]
The draw order of the stroke/fill is intentional. The canvas text rendering does not allow us to specify inside/outside/center for the stroke alignment. We made a decision to put the stroke behind to emulate outside stroke alignment, which is more coming styling. I am not sure we are doubling the stroke to account for this so 1pm may look weird with subpixel rendering. A style option to specify draw order is a good idea!

**Comment by maybemaynot** [comment]
not only 1px weird for my case
no problem about subpixel rendering
but i think that maybe we can do sothing to make line looks better
(eg: the top same to bottom, the left same to right, even if they all subpixel render)

look at image bellow, only 3char fontsize= 201,204,207 looks right

![image](https://user-images.githubusercontent.com/88305215/215428873-2ea9f305-d4f1-4293-8185-7ef157b10b4f.png)

**Comment by bigtimebuddy** [comment]
Here's an example of how canvas (no PixiJS) renders the outlines on text:
https://jsfiddle.net/bigtimebuddy/pz4rcdf6/

You can see the same behavior. It happens when strokes don't fall on an even pixels. They blurry and it's looks terrible.

If you care deeply about this, my suggestion is that you create a bitmap-font and use BitmapText instead. That's the only way to ensure pixel accuracy with strokes.

**Comment by maybemaynot** [comment]
@bigtimebuddy 

yeap, your are right !
I have make it work by two Text

one with only fill but no stroke
another one with only stroke but non fill

this make my thin border line looks better

**Comment by maybemaynot** [comment]
> 

I got this may help ~~~

```
  // center
  context.fillText("Text", 640, 360);
  context.strokeText("Text", 640, 360);

  // outside
  context.strokeText("Text", 640, 360);
  context.fillText("Text", 640, 360);

  // inside
  context.fillText("Text", 640, 360);
  context.strokeText("Text", 640, 360);
  context.globalCompositeOperation = "destination-in";
  context.fillStyle = "#000";
  context.fillText("Text", 640, 360);
```

**Comment by maybemaynot** [comment]
and more

```
  // fill only
  context.fillText("Text", 640, 360);

  // fill with center stroke
  context.fillText("Text", 640, 360);
  context.strokeText("Text", 640, 360);

  // fill with outside stroke
  context.strokeText("Text", 640, 360);
  context.fillText("Text", 640, 360);

  // fill with inside stroke
  context.fillText("Text", 640, 360);
  context.strokeText("Text", 640, 360);
  context.globalCompositeOperation = "destination-in";
  context.fillStyle = "#fff";
  context.fillText("Text", 640, 360);

  // fill with outside stroke(transparent)
  // this case is same to fill only

  // fill with center/inside stroke(transparent)
  // without this code, a thin line will be see
  context.strokeText("Text", 640, 360);
  context.fillText("Text", 640, 360);
  context.globalCompositeOperation = "xor";
  // single char fine
  // but not fine when lineWidth bigger than distance of two chars,
  // eg: Text, dirty part bwtween x&t
  context.strokeText("Text", 640, 360);
  // this can fix the dirty as above
  context.globalCompositeOperation = "destination-in";
  context.fillStyle = "#fff";
  // context.lineWidth -= 10;
  context.fillText("Text", 640, 360);

  // stroke only, center
  context.strokeText("Text", 640, 360);

  // stroke only, outside
  context.strokeText("Text", 640, 360);
  context.globalCompositeOperation = "destination-out";
  context.fillStyle = "#fff";
  context.fillText("Text", 640, 360);

  // stroke only, inside
  context.strokeText("Text", 640, 360);
  context.globalCompositeOperation = "destination-in";
  context.fillStyle = "#fff";
  context.fillText("Text", 640, 360);
```

but there are tow **important** things

1. this could be effect dropshadow
2. another thing, should be re measured and resize the canvas?

**Comment by bigtimebuddy** [comment]
I'm not clear on your questions. Can you rephrase?

**Comment by maybemaynot** [comment]
> I'm not clear on your questions. Can you rephrase?

which one?
